Remote Static Call

puma314 opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

A draft PR implementing the RFP described here: https://github.com/ethereum-optimism/ecosystem-contributions/issues/76

**Tests**

Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.

**Additional context**

Add any other context about the problem you're solving.

**Metadata**

- Fixes #[Link to Issue]

puma314 committed:
Started adding remote static call

puma314 committed:
Thread through precompile info to RunPrecompiledContract

puma314 committed:
Got precompile contract test working

technicallykind commented:
👀 looking good @puma314

tynes commented on cmd/utils/flags.go:
```diff
@@ -289,6 +289,11 @@ var (
 		Category: flags.EthCategory,
 	}
+	OverrideL1ArchiveNodeRPC = &cli.StringFlag{
+		Name:     "override.l1archivenoderpc",
```
The `override` namespace is for flags that override values hardcoded in the software, this doesn't necessarily feel like it belongs there. There is a flag for the sequencer RPC url, it should be defined with a name similar to that

tynes commented on core/types/rollup_l1_cost.go:
```diff
@@ -45,10 +45,13 @@ type StateGetter interface {
 type L1CostFunc func(blockNum uint64, blockTime uint64, dataGas RollupGasData, isDepositTx bool) *big.Int
 
+// TODO: are these slots updated / correct?
+// https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/L1Block.sol#L39C28-L39C28
```
These haven't changed, solidity packs storage slots automagically when it can, not all values are `uint256`

puma314 replied:
I'm silly, that makes sense

tynes commented on core/vm/contracts.go:
```diff
@@ -1048,3 +1064,72 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
+		]
+	}]`
+
+	parsedABI, err := abi.JSON(strings.NewReader(ethcallAbi))
```
Ideally we can manually parse the data from the bytes directly but this is great for a first pass

puma314 replied:
Done

tynes replied:
Maybe we want to remove ABI completely from consensus

```
[0:12] - to, the address
[12:16] - uint32 data length
[14:] - data
```

The logic could be pretty simple, assert the min length, copy bytes into an address, read the length into a uint32, check that the calldata is long enough, copy the calldata then make the remote call then pass back the return data

tynes replied:
The call could be created with the following library pseudocode:

```solidity
library RemoteStaticCall {
    function call(address addr, bytes memory data) internal returns (bool success, bytes memory returndata) {
        (success, returndata) = address(0xXX).call(abi.encodePacked(addr, uint32(data.length), data));
    }
}
```

I think its important to have a nice high level abstraction for devs to use

puma314 replied:
Oh I didn't realize the motivation was to not use the `abi` library in `op-geth` at all, but that actually makes a ton of sense as to why you would want to just decode the bytes in a raw way. I changed this to just do the decoding manually since it's not that complicated for `bytes memory` anyways.

tynes commented on core/vm/contracts.go:
```diff
@@ -1048,3 +1064,72 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
+	// ABI definition to decode `input` data
+	const ethcallAbi = `[{
+		"type" : "function",
+		"name" : "eth_call",
```
Ideally the interface to the precompile is `abi.encode(address, bytes)` so we don't need to worry about the 4byte selector being prefixed. We could write a simple solidity library that wraps the abi encode and returns the bytes or even a library that makes the call after abi encoding the args so that devs have a nice interface.

If we wanted to introduce additional functionality like other sorts of calls like fetching L1 balance we could introduce it in a new contract instead of having a precompile with a dispatch table in it based on the 4byte selector

puma314 replied:
Ok I changed it to just be `abi.encode(address, bytes)`, although fwiw I kind of like the idea of having dispatch based on 4 byte selector all to the same address

tynes commented on params/config.go:
```diff
@@ -309,9 +309,9 @@ type ChainConfig struct {
-
+	BedrockBlock     *big.Int `json:"bedrockBlock,omitempty"`     // Bedrock switch block (nil = no fork, 0 = already on optimism bedrock)
+	RegolithTime     *uint64  `json:"regolithTime,omitempty"`     // Regolith switch time (nil = no fork, 0 = already on optimism regolith)
+	L1ArchiveNodeRPC *string  `json:"l1ArchiveNodeRPC,omitempty"` // L1 Archive Node RPC URL (nil = no fork, 0 = already on optimism l1 archive node)
```
Hmm this doesn't feel like the correct place necessarily because this is all of the hardfork info

puma314 replied:
Regarding this comment, do you think the `Optimism *OptimismConfig `json:"optimism,omitempty"` would be a better place for it?

tynes replied:
It feels like its best for it to just be a CLI flag, these config structs are written to disk as part of the genesis info, I don't think it needs to be there

tynes commented on cmd/utils/flags.go:
```diff
@@ -289,6 +289,11 @@ var (
 	}
+	OverrideL1ArchiveNodeRPC = &cli.StringFlag{
+		Name:     "override.l1archivenoderpc",
+		Usage:    "Manually specify the L1 Archive Node RPC URL, overriding the bundled setting",
```
We should make this required so that the software doesn't start without it, otherwise there will be a fork chain that people commonly end up on

puma314 commented on core/vm/contracts.go:
```diff
@@ -79,15 +94,16 @@ var PrecompiledContractsIstanbul = map[common.Address]PrecompiledContract{
-	common.BytesToAddress([]byte{7}): &bn256ScalarMulIstanbul{},
-	common.BytesToAddress([]byte{8}): &bn256PairingIstanbul{},
-	common.BytesToAddress([]byte{9}): &blake2F{},
+	common.BytesToAddress([]byte{1}):  &ecrecover{},
```
Do you guys have special formatter settings, I'm not sure why this whole thing is showing up as a diff?

tynes replied:
I believe we just use standard `gofmt` or `goimports`, this diff is from 2 whitespace between the addr and precompile structs, prob autogen'd by `gofmt`

puma314 committed:
Comments

puma314 committed:
Further cleanup

puma314 committed:
Fix test

tynes commented on core/vm/contracts.go:
```diff
@@ -1048,3 +1062,65 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
+		return nil, errors.New("L1 archive RPC is not configured")
+	}
+
+	rpcCtx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
```
This is a bit risky, if a single request times out then you'll end up on a fork. We probably don't want a timeout since we trust the remote node anyways, preferring safety over liveness

puma314 replied:
Ok sounds good, will just remove timeout

tynes commented on core/genesis.go:
```diff
@@ -272,6 +272,7 @@ type ChainOverrides struct {
 	OverrideOptimismBedrock  *big.Int
 	OverrideOptimismRegolith *uint64
 	OverrideOptimism         *bool
+	L1ArchiveNodeRPC         *string
```
Is this dead code?

puma314 replied:
yes, good catch ,removed

tynes commented on core/vm/contracts.go:
```diff
@@ -1048,3 +1062,65 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
+type remoteStaticCall struct{}
+
+func (c *remoteStaticCall) RequiredGas(input []byte) uint64 {
+	return params.RemoteStaticCallGas
```
If we move to the non abi format, then we can have a function that charges based on the size of the data since it'll be easy to pull out the size, this is ok for now but def would be good to include when it comes to the gas metering thoughts

puma314 replied:
Yeah we can definitely do that, something like `params.RemoteStaticCallGasMultiplier * len(input)` or something like that?

puma314 replied:
As an update, I just changed it to that: `params.RemoteStaticCallGas * uint64(len(data))`

puma314 committed:
Added comments and removed abi

puma314 committed:
Refactored required gas as well

tynes committed:
build: fix

```
./build/bin/geth --http --http.api web3,eth,debug,personal,net --vmdebug --dev --allow-insecure-unlock --rollup.l1archivenoderpc $ETH_RPC_URL
```

puma314 committed:
Not required

adraffy commented:
just saw Vitalik mention this on X

I never heard of `REMOTESTATICCALL` before but this seems incredibly useful!
